btree/btree.c

```diff
@@ -1885,12 +1869,45 @@ static int64_t mem_tree_get(struct node_op *op, struct segaddr *addr, int nxt);
  */
 static int64_t tree_get(struct node_op *op, struct segaddr *addr, int nxt)
 {
-	int        nxt_state;
+	struct td              *tree = NULL;
+	struct nd              *node = NULL;
+
+	if (addr != NULL) {
+		nxt  = node_get(op, NULL, addr, nxt);
+		node = op->no_node;
+		if (node != NULL) {
+			tree = node->n_tree;
+		} else
+			op->no_op.o_sm.sm_rc = M0_ERR(-EINVAL);
+
+		if (tree == NULL) {
+			tree = m0_alloc(sizeof *tree);
+			m0_rwlock_init(&tree->t_lock);
+			m0_rwlock_write_lock(&tree->t_lock);
 
-	nxt_state = segops->so_tree_get(op, addr, nxt);
+			tree->t_ref = 1;
+			tree->t_starttime = 0;
+			tree->t_root = node;
 
-	return nxt_state;
+			node_lock(node);
+			node->n_tree = tree;
+			node_unlock(node);
+
+		} else {
+			m0_rwlock_write_lock(&tree->t_lock);
+			tree->t_ref++;
+		}
+
+		op->no_node = tree->t_root;
+		op->no_tree = tree;
+		m0_rwlock_write_unlock(&tree->t_lock);
+
+	} else
+		op->no_op.o_sm.sm_rc = M0_ERR(-EINVAL);
```

Comment:
Are we assuming that a that tree_create will always be called with a valid `addr` ?

Reply:
Currently, yes. I have assumed that. I need to see if we have a check to determine this.

Reply:
Ok.